hw/ip/otbn/dv/rig/rig/gens/untaken_branch.py

```diff
@@ -0,0 +1,94 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import random
+from typing import Optional
+
+from shared.insn_yaml import InsnsFile
+
+from ..config import Config
+from ..model import Model
+from ..program import ProgInsn, Program
+
+from ..snippet import ProgSnippet
+from ..snippet_gen import GenCont, GenRet
+from .branch_gen import BranchGen
+
+
+class UntakenBranch(BranchGen):
+    '''A snippet generator to generate aligned or
+       unalligned addresses for which branches are not taken.
+    '''
+
+    def __init__(self, cfg: Config, insns_file: InsnsFile) -> None:
+        super().__init__(cfg, insns_file)
+
+    def gen(self,
+            cont: GenCont,
+            model: Model,
+            program: Program) -> Optional[GenRet]:
+
+        # Return None if this is the last instruction in the current gap
+        # because we need to either jump or do an ECALL to avoid getting stuck.
+        if program.get_insn_space_at(model.pc) <= 1:
+            return None
+
+        # Pick maximum and minimum address range for the current PC.
+        imm_rng = self.imm_op_type.get_op_val_range(model.pc)
+        assert imm_rng is not None
+
+        min_addr, max_addr = imm_rng
+
+        # Get known registers. We always have x0.
+        # So it should never fail.
+        known_regs = model.regs_with_known_vals('gpr')
+        assert known_regs
+
+        # Pick a random register among known registers.
+        idx, value = random.choice(known_regs)
+
+        equals = []
+        not_equals = []
+
+        for reg_idx, reg_val in known_regs:
+            if reg_val == value:
+                equals.append(reg_idx)
+            else:
+                not_equals.append(reg_idx)
```

Comment:
This loop is the same in `BadBranch`: maybe worth lifting to `BranchGen`?

Reply:
I have made the necessary changes. 
BadBranch and UntakenBranches have three new functions instead of gen function: 
1. chosen_insn
2. chosen_insn
3. snippet_end